 <img src="https://github.com/MahendraCherukupalli/The-Sparks-Foundation-Internship/blob/master/wpl_badge1.png?raw=true"/>

## Steps:
- Import Libraries
- Get the Data
- Data Preprocessing
- Train-Test Split
- Scalling the Data
- Creating and Training the Model
- Predictions
- Residual Analysis
- Evaluating the modelm

**Let's get started!**

# Import Libraries

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns  
import matplotlib.pyplot as plt  
%matplotlib inline

# Get the data

In [3]:
data=pd.read_csv("matches.csv")
data.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

In [4]:
data.shape

(756, 18)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
id                 756 non-null int64
season             756 non-null int64
city               749 non-null object
date               756 non-null object
team1              756 non-null object
team2              756 non-null object
toss_winner        756 non-null object
toss_decision      756 non-null object
result             756 non-null object
dl_applied         756 non-null int64
winner             752 non-null object
win_by_runs        756 non-null int64
win_by_wickets     756 non-null int64
player_of_match    752 non-null object
venue              756 non-null object
umpire1            754 non-null object
umpire2            754 non-null object
umpire3            119 non-null object
dtypes: int64(5), object(13)
memory usage: 106.4+ KB


In [6]:
data.describe()

id       season  dl_applied  win_by_runs  win_by_wickets
count    756.000000   756.000000  756.000000   756.000000      756.000000
mean    1792.178571  2013.444444    0.025132    13.283069        3.350529
std     3464.478148     3.366895    0.156630    23.471144        3.387963
min        1.000000  2008.000000    0.000000     0.000000        0.000000
25%      189.750000  2011.000000    0.000000     0.000000        0.000000
50%      378.500000  2013.000000    0.000000     0.000000        4.000000
75%      567.250000  2016.000000    0.000000    19.000000        6.000000
max    11415.000000  2019.000000    1.000000   146.000000       10.000000

In [7]:
data1=pd.read_csv("deliveries.csv")
data1.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  ...  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0  ...         0   
1     2  DA Warner    S Dhawan  TS Mills              0  ...         0   
2     3  DA Warner    S Dhawan  TS Mills              0  ...         0   
3     4  DA Warner    S Dhawan  TS Mills              0  ...         0   
4     5  DA Warner    S Dhawan  TS Mills              0  ...         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs  player_dismissed dismissal_kind fielder  
0           0               NaN            NaN     NaN  
1           0               NaN            NaN     NaN  
2           4               NaN            NaN     NaN  
3           0               NaN            NaN     NaN  
4           2               NaN            NaN     NaN  

[5 rows x 21 columns]

In [8]:
data1.shape

(179078, 21)

In [9]:
data1.describe()

match_id         inning           over           ball  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean     1802.252957       1.482952      10.162488       3.615587   
std      3472.322805       0.502074       5.677684       1.806966   
min         1.000000       1.000000       1.000000       1.000000   
25%       190.000000       1.000000       5.000000       2.000000   
50%       379.000000       1.000000      10.000000       4.000000   
75%       567.000000       2.000000      15.000000       5.000000   
max     11415.000000       5.000000      20.000000       9.000000   

       is_super_over      wide_runs       bye_runs    legbye_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.000452       0.036721       0.004936       0.021136   
std         0.021263       0.251161       0.116480       0.194908   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       5.000000       4.000000       5.000000   

         noball_runs   penalty_runs   batsman_runs     extra_runs  \
count  179078.000000  179078.000000  179078.000000  179078.000000   
mean        0.004183       0.000056       1.246864       0.067032   
std         0.070492       0.016709       1.608270       0.342553   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       1.000000       0.000000   
75%         0.000000       0.000000       1.000000       0.000000   
max         5.000000       5.000000       7.000000       7.000000   

          total_runs  
count  179078.000000  
mean        1.313897  
std         1.605422  
min         0.000000  
25%         0.000000  
50%         1.000000  
75%         1.000000  
max        10.000000

# Data Preprocessing

In [10]:
# checking for null values...
data.isnull().sum()

id                   0
season               0
city                 7
date                 0
team1                0
team2                0
toss_winner          0
toss_decision        0
result               0
dl_applied           0
winner               4
win_by_runs          0
win_by_wickets       0
player_of_match      4
venue                0
umpire1              2
umpire2              2
umpire3            637
dtype: int64

In [11]:
data1.isnull().sum()

match_id                 0
inning                   0
batting_team             0
bowling_team             0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
is_super_over            0
wide_runs                0
bye_runs                 0
legbye_runs              0
noball_runs              0
penalty_runs             0
batsman_runs             0
extra_runs               0
total_runs               0
player_dismissed    170244
dismissal_kind      170244
fielder             172630
dtype: int64

In [12]:
MI=data["team1"]=='Mumbai Indians'
DC=data["team1"]=='Delhi Daredevils'
MI1=data["team2"]=='Mumbai Indians'
DC1=data["team2"]=='Delhi Daredevils'

In [13]:
rdf=data[(MI & DC1)|(MI1 & DC)]

In [14]:
rdf.head()

id  season         city        date           team1             team2  \
23    24    2017       Mumbai  2017-04-22  Mumbai Indians  Delhi Daredevils   
43    44    2017        Delhi  2017-05-06  Mumbai Indians  Delhi Daredevils   
81    82    2008       Mumbai  2008-05-04  Mumbai Indians  Delhi Daredevils   
108  109    2008        Delhi  2008-05-24  Mumbai Indians  Delhi Daredevils   
149  150    2009  East London  2009-05-08  Mumbai Indians  Delhi Daredevils   

          toss_winner toss_decision  result  dl_applied            winner  \
23   Delhi Daredevils         field  normal           0    Mumbai Indians   
43   Delhi Daredevils         field  normal           0    Mumbai Indians   
81   Delhi Daredevils         field  normal           0    Mumbai Indians   
108  Delhi Daredevils         field  normal           0  Delhi Daredevils   
149    Mumbai Indians           bat  normal           0  Delhi Daredevils   

     win_by_runs  win_by_wickets player_of_match                       venue  \
23            14               0  MJ McClenaghan            Wankhede Stadium   
43           146               0     LMP Simmons            Feroz Shah Kotla   
81            29               0      SM Pollock  Dr DY Patil Sports Academy   
108            0               5      KD Karthik            Feroz Shah Kotla   
149            0               7         A Nehra                Buffalo Park   

            umpire1      umpire2 umpire3  
23   A Nand Kishore       S Ravi     NaN  
43      Nitin Menon    CK Nandan     NaN  
81        IL Howell  RE Koertzen     NaN  
108       BF Bowden  K Hariharan     NaN  
149       M Erasmus  SK Tarapore     NaN

In [15]:
rdf[["team1","team2","toss_winner","toss_decision","winner"]]

team1             team2       toss_winner toss_decision  \
23     Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
43     Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
81     Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
108    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
149    Mumbai Indians  Delhi Daredevils    Mumbai Indians           bat   
169    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
182    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
220    Mumbai Indians  Delhi Daredevils    Mumbai Indians           bat   
237  Delhi Daredevils    Mumbai Indians  Delhi Daredevils           bat   
281    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
326    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
341  Delhi Daredevils    Mumbai Indians    Mumbai Indians         field   
389    Mumbai Indians  Delhi Daredevils    Mumbai Indians           bat   
407    Mumbai Indians  Delhi Daredevils    Mumbai Indians           bat   
472    Mumbai Indians  Delhi Daredevils    Mumbai Indians           bat   
507    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
538  Delhi Daredevils    Mumbai Indians    Mumbai Indians         field   
555  Delhi Daredevils    Mumbai Indians  Delhi Daredevils           bat   
592  Delhi Daredevils    Mumbai Indians    Mumbai Indians         field   
622    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
644    Mumbai Indians  Delhi Daredevils  Delhi Daredevils         field   
690  Delhi Daredevils    Mumbai Indians  Delhi Daredevils           bat   

               winner  
23     Mumbai Indians  
43     Mumbai Indians  
81     Mumbai Indians  
108  Delhi Daredevils  
149  Delhi Daredevils  
169  Delhi Daredevils  
182    Mumbai Indians  
220    Mumbai Indians  
237    Mumbai Indians  
281    Mumbai Indians  
326  Delhi Daredevils  
341  Delhi Daredevils  
389    Mumbai Indians  
407  Delhi Daredevils  
472  Delhi Daredevils  
507    Mumbai Indians  
538  Delhi Daredevils  
555    Mumbai Indians  
592  Delhi Daredevils  
622    Mumbai Indians  
644  Delhi Daredevils  
690  Delhi Daredevils

In [16]:
rdf[["team1","team2","toss_winner","toss_decision","winner"]].isnull().sum()

team1            0
team2            0
toss_winner      0
toss_decision    0
winner           0
dtype: int64

# 1

In [42]:
rdf[(rdf["toss_winner"]=="Delhi Daredevils")&(rdf["toss_decision"]=="field")]

id  season           city        date           team1  \
23     24    2017         Mumbai  2017-04-22  Mumbai Indians   
43     44    2017          Delhi  2017-05-06  Mumbai Indians   
81     82    2008         Mumbai  2008-05-04  Mumbai Indians   
108   109    2008          Delhi  2008-05-24  Mumbai Indians   
169   170    2009      Centurion  2009-05-21  Mumbai Indians   
182   183    2010          Delhi  2010-03-17  Mumbai Indians   
281   282    2011         Mumbai  2011-05-07  Mumbai Indians   
326   327    2012         Mumbai  2012-04-16  Mumbai Indians   
507   508    2014         Mumbai  2014-05-23  Mumbai Indians   
622   623    2016  Visakhapatnam  2016-05-15  Mumbai Indians   
644  7902    2018         Mumbai    14/04/18  Mumbai Indians   

                team2       toss_winner toss_decision  result  dl_applied  \
23   Delhi Daredevils  Delhi Daredevils         field  normal           0   
43   Delhi Daredevils  Delhi Daredevils         field  normal           0   
81   Delhi Daredevils  Delhi Daredevils         field  normal           0   
108  Delhi Daredevils  Delhi Daredevils         field  normal           0   
169  Delhi Daredevils  Delhi Daredevils         field  normal           0   
182  Delhi Daredevils  Delhi Daredevils         field  normal           0   
281  Delhi Daredevils  Delhi Daredevils         field  normal           0   
326  Delhi Daredevils  Delhi Daredevils         field  normal           0   
507  Delhi Daredevils  Delhi Daredevils         field  normal           0   
622  Delhi Daredevils  Delhi Daredevils         field  normal           0   
644  Delhi Daredevils  Delhi Daredevils         field  normal           0   

               winner  win_by_runs  win_by_wickets player_of_match  \
23     Mumbai Indians           14               0  MJ McClenaghan   
43     Mumbai Indians          146               0     LMP Simmons   
81     Mumbai Indians           29               0      SM Pollock   
108  Delhi Daredevils            0               5      KD Karthik   
169  Delhi Daredevils            0               4        V Sehwag   
182    Mumbai Indians           98               0    SR Tendulkar   
281    Mumbai Indians           32               0       AT Rayudu   
326  Delhi Daredevils            0               7        S Nadeem   
507    Mumbai Indians           15               0      MEK Hussey   
622    Mumbai Indians           80               0       KH Pandya   
644  Delhi Daredevils            0               7          JJ Roy   

                                                 venue               umpire1  \
23                                    Wankhede Stadium        A Nand Kishore   
43                                    Feroz Shah Kotla           Nitin Menon   
81                          Dr DY Patil Sports Academy             IL Howell   
108                                   Feroz Shah Kotla             BF Bowden   
169                                    SuperSport Park             IL Howell   
182                                   Feroz Shah Kotla           BR Doctrove   
281                                   Wankhede Stadium           K Hariharan   
326                                   Wankhede Stadium             BF Bowden   
507                                   Wankhede Stadium                S Ravi   
622  Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket St...           Nitin Menon   
644                                   Wankhede Stadium  K Ananthapadmanabhan   

         umpire2     umpire3  
23        S Ravi         NaN  
43     CK Nandan         NaN  
81   RE Koertzen         NaN  
108  K Hariharan         NaN  
169       S Ravi         NaN  
182  SK Tarapore         NaN  
281   SJA Taufel         NaN  
326  SK Tarapore         NaN  
507    RJ Tucker         NaN  
622    CK Nandan         NaN  
644  Nitin Menon  Rod Tucker

In [43]:
rdf["winner"].value_counts()

Delhi Daredevils    11
Mumbai Indians      11
Name: winner, dtype: int64

according to the dataset its TIE...

In [19]:
data.columns

Index(['id', 'season', 'city', 'date', 'team1', 'team2', 'toss_winner',
       'toss_decision', 'result', 'dl_applied', 'winner', 'win_by_runs',
       'win_by_wickets', 'player_of_match', 'venue', 'umpire1', 'umpire2',
       'umpire3'],
      dtype='object')

In [20]:
data1.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

# 2

In [21]:
wickets=data1[(data1["batting_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))]

In [22]:
wickets1=wickets[wickets["over"].isin([1,2,3,4,5,6])]

In [23]:
wickets1.groupby("match_id")["player_dismissed"].count()

match_id
24       7
44       5
82       4
109      2
150      3
170      4
183      2
221      2
238      4
282      5
327      3
342      3
390      4
408      2
473      2
508      1
539      3
556      6
593      2
623      2
7902     1
7948     3
11139    5
11325    0
Name: player_dismissed, dtype: int64

In [24]:
wickets1.groupby(["match_id"])["player_dismissed"].count().mean()

3.125

According to the dataset both teams loss average of 3 wickets in powerplay

# 3

In [25]:
ecomony_rate=data1[(data1["batting_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))]

df=pd.DataFrame()

df["balls_bowled"]=ecomony_rate.groupby(["match_id","bowler"])["ball"].count()

df["given_runs"]=ecomony_rate.groupby(["match_id","bowler"])["total_runs"].sum()

ecomony_rate0=ecomony_rate[ecomony_rate["total_runs"].isin([4,6])]

df["boundaries"]=ecomony_rate0.groupby(["match_id","bowler"])["total_runs"].count()

ecomony_rate1=ecomony_rate[ecomony_rate["total_runs"]==4]

df["boundaries_4"]=ecomony_rate1.groupby(["match_id","bowler"])["total_runs"].count()

ecomony_rate2=ecomony_rate[ecomony_rate["total_runs"]==6]

df["boundaries_6"]=ecomony_rate2.groupby(["match_id","bowler"])["total_runs"].count()

df=df.fillna(0)

df

balls_bowled  given_runs  boundaries  boundaries_4  \
match_id bowler                                                                
24       A Mishra                   25          19         2.0           2.0   
         CH Morris                  25          33         4.0           2.0   
         HH Pandya                  20          25         2.0           2.0   
         Harbhajan Singh            24          26         3.0           2.0   
         JJ Bumrah                  25          21         2.0           2.0   
...                                ...         ...         ...           ...   
11325    K Paul                     19          38         4.0           2.0   
         K Rabada                   26          42         6.0           5.0   
         KH Pandya                  12           7         0.0           0.0   
         RD Chahar                  26          21         2.0           2.0   
         SL Malinga                 25          46         4.0           3.0   

                          boundaries_6  
match_id bowler                         
24       A Mishra                  0.0  
         CH Morris                 2.0  
         HH Pandya                 0.0  
         Harbhajan Singh           1.0  
         JJ Bumrah                 0.0  
...                                ...  
11325    K Paul                    2.0  
         K Rabada                  1.0  
         KH Pandya                 0.0  
         RD Chahar                 0.0  
         SL Malinga                1.0  

[283 rows x 5 columns]

In [26]:
def status(a,b,c,d):
    if a>7:
        if b<=6:
            if c==2:
                economy=(d/a)
                if economy<=9:
                    return "Yes"
                else:
                    return "No"
            else:
                return "No"
        else:
            return "No"
    else:
        return "No"


In [27]:
df["status"]=df.apply(lambda x: status(x.balls_bowled, x.boundaries,x.boundaries_6,x.given_runs), axis=1)

In [28]:
df

balls_bowled  given_runs  boundaries  boundaries_4  \
match_id bowler                                                                
24       A Mishra                   25          19         2.0           2.0   
         CH Morris                  25          33         4.0           2.0   
         HH Pandya                  20          25         2.0           2.0   
         Harbhajan Singh            24          26         3.0           2.0   
         JJ Bumrah                  25          21         2.0           2.0   
...                                ...         ...         ...           ...   
11325    K Paul                     19          38         4.0           2.0   
         K Rabada                   26          42         6.0           5.0   
         KH Pandya                  12           7         0.0           0.0   
         RD Chahar                  26          21         2.0           2.0   
         SL Malinga                 25          46         4.0           3.0   

                          boundaries_6 status  
match_id bowler                                
24       A Mishra                  0.0     No  
         CH Morris                 2.0    Yes  
         HH Pandya                 0.0     No  
         Harbhajan Singh           1.0     No  
         JJ Bumrah                 0.0     No  
...                                ...    ...  
11325    K Paul                    2.0    Yes  
         K Rabada                  1.0     No  
         KH Pandya                 0.0     No  
         RD Chahar                 0.0     No  
         SL Malinga                1.0     No  

[283 rows x 6 columns]

In [29]:
df.groupby("match_id")["status"].value_counts()

match_id  status
24        No         9
          Yes        2
44        No        10
          Yes        2
82        No        14
109       No         9
          Yes        2
150       No        11
          Yes        1
170       No        11
          Yes        1
183       No         9
          Yes        2
221       No        12
238       No        12
282       No        11
327       No        10
342       No        10
          Yes        1
390       No         9
          Yes        2
408       No        10
          Yes        2
473       No        11
          Yes        1
508       No        10
          Yes        2
539       No         9
          Yes        4
556       No         8
          Yes        4
593       No        12
          Yes        1
623       No         9
          Yes        2
7902      No        11
          Yes        1
7948      No         6
          Yes        6
11139     No        11
          Yes        1
11325     No        10
          Yes    

According to the dataset 0-2 bowlers satisfied the given condition

# 4

In [30]:
more_30=data1[(data1["batting_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"].isin(["Mumbai Indians","Delhi Daredevils","Delhi Capitals"]))]

In [31]:
df1=pd.DataFrame(more_30.groupby(["match_id","batsman"])["batsman_runs"].sum())

In [32]:
df2=df1["batsman_runs"]>=30

In [33]:
df1[df2]

batsman_runs
match_id batsman                  
24       CH Morris              52
         K Rabada               44
44       KA Pollard             63
         LMP Simmons            66
82       DJ Thornely            30
...                            ...
11325    HH Pandya              32
         KH Pandya              41
         Q de Kock              36
         RG Sharma              31
         S Dhawan               35

[89 rows x 1 columns]

According to the dataset 3-4 batsmens score more than 30 runs

# 5

In [34]:
Bumrah=data1[(data1["batting_team"].isin(["Delhi Daredevils","Delhi Capitals"]))&(data1["bowling_team"]=="Mumbai Indians")&(data1["bowler"]=="JJ Bumrah")]

In [35]:
Bumrah1=Bumrah[Bumrah["over"].isin([16,17,18,19,20])]

In [36]:
df3=pd.DataFrame()

In [37]:
df3["balls"]=Bumrah1.groupby("match_id")["ball"].count()

In [38]:
df3["runs_given"]=Bumrah1.groupby("match_id")["total_runs"].sum()

In [39]:
df3["economy"]=df3["runs_given"]/(df3["balls"]/6)

In [40]:
df3

balls  runs_given  economy
match_id                            
24           12           9     4.50
408           8          13     9.75
508          12          21    10.50
539          12          25    12.50
593          20          35    10.50
623          12           6     3.00
7902         12          16     8.00
7948          8          10     7.50
11139        12          32    16.00
11325        12          13     6.50

In [41]:
df3["economy"].mean()

8.875

According to the dataset bumrah economy rate in death overs is around 8.9